ligo/haskell/src/Ligo/Typescript.hs

```diff
@@ -0,0 +1,380 @@
+toEpType (NTLambda _ f1 f2) = EpLambda (toEpType f1) (toEpType f2)
+toEpType (NTBytes _) = EpBytes
+toEpType (NTMutez _) = EpMutez
+toEpType (NTKeyHash _) = EpKeyHash
+toEpType (NTTimestamp _) = EpTimestamp
+toEpType (NTAddress _) = EpAddress
+toEpType (NTBool _) = EpBool
+
+-- looks up the entrypoints in the Notes t for the whole parameter.
+-- It might be possible to do this without re-implementing logic in
+-- the `mkEntrypointsMap` from morley.
+lookupEntryPointsAndTypes :: Notes t -> [Entrypoint]
+lookupEntryPointsAndTypes (NTOr _ fa1 fa2 f1 f2) =
+  let
+    leftEntrypoint = case extractAnn fa1 of
+      Just ann_ -> (Entrypoint ann_ (toEpType f1)) : (lookupEntryPointsAndTypes f1)
+      Nothing -> lookupEntryPointsAndTypes f1
+    rightEntrypoint = case extractAnn fa2 of
+      Just ann_ -> (Entrypoint ann_ (toEpType f2)) : (lookupEntryPointsAndTypes f2)
+      Nothing -> lookupEntryPointsAndTypes f2
+  in leftEntrypoint <> rightEntrypoint
+lookupEntryPointsAndTypes _ = []
+
+-- Check if the annotation in this annotation is among the
+-- ones we are not interested in, or empty. If so return a Nothing. Else
+-- wrap the annotation in a Just before returning it.
+extractAnn :: Annotation tag -> Maybe Text
+extractAnn ann_ =
+  case (ann_ == noAnn) || (isJust $ find (== ann_) ignoreEntypoints) of
+    True -> Nothing
+    _ -> Just $ unAnnotation ann_
+
+-- Some entrypoints we don't want to call directly.
+ignoreEntypoints :: [Annotation tag]
+ignoreEntypoints = [ann "running", ann "transfer_contract_tokens", ann "xtzForbidden", ann "call_FA2", ann "migratable"]
+
+type Typename = Text
+type Fieldname = Text
+
+-- Represents a Typescript declaration. For now only contains one that
+-- defines a Type.
+data TsDecl
+  = TsType Text TsTypeDef
+  deriving stock Generic
+
+-- Represents an instance of a Typescript type.
+data TsType
+  = TsNumber
+  | TsString
+  | TsBool
+  | TsUnit
+  | TsOption TsType
+  | TsList TsType
+  | TsMap TsType TsType
+  | TsBigMap TsType TsType
+  | TsCustom Typename
+  deriving stock Generic
+
+-- Implements the conversion of Type to its source code representation.
+instance Buildable TsType where
+  build = \case
+    TsNumber -> "number"
+    TsString -> "string"
+    TsBool -> "boolean"
+    TsUnit -> "void"
+    TsOption t_ -> "null | " +| (build t_)
+    TsList t_ -> "Array<" +| t_ |+ ">"
+    TsMap k v -> "Map<" +| k |+ "," +| v |+ ">"
+    TsBigMap k v -> "Map<" +| k |+ "," +| v |+ ">"
+    TsCustom tn -> build tn
+
+-- Represents a Typescript module with just enough stuff we are interested.
+data TsModule = TsModule
+  { tsmEpName :: Text
+  , tsmImports :: [Text]
+  , tsmDecls :: [TsDecl]
+  } deriving stock Generic
+
+instance Buildable TsModule where
+  build (TsModule name imports decls) = Fmt.unlinesF [build name, build imports, build decls]
+
+type IsOptional = Bool
+
+-- Represents the right side of a Type definition in Typescript.
+data TsTypeDef
+  = TsSum [Typename]
```

Comment:
I think `|` and `&` are called union and intersection types in TypeScript (rather than sum and product types like in Haskell)

Reply:
Have renamed `TsSum` to `TsUnion`.